### PR TITLE
Trans lab modified

### PR DESCRIPTION
**Pull request by misha-kis**

### CODE REVIEW

File: detectors/TransLab/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM python:3.9-bullseye
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/Dockerfile]: Add GPU support please

File: detectors/TransLab/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+WORKDIR /detector
+
+COPY . .
+
+RUN apt-get update && apt-get install ffmpeg libsm6 libxext6  -y \
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/Dockerfile]: Separate install requirements for python from apt requirements. Make a separate layer for apt requirements

Comment (kirill-ivanov-a) [reply detectors/TransLab/Dockerfile]: Also a separate layer for downloading weights for the neural network and installing the application itself

File: detectors/TransLab/common/network_runner_base.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#  @Date    : 2023-03-19
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/common/network_runner_base.py]: Take the adapter common to all algorithms to a higher level so as not to copy

File: detectors/TransLab/README.md [hunk 4]
```diff
@@ -19,6 +19,10 @@ For downloading the data, you can refer to [Trans10K Website](https://xieenze.gi
 ```
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/README.md:15]: Add info about docker build

File: detectors/TransLab/network_runner.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
+from __future__ import print_function
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/network_runner.py:15]: Add license

File: detectors/TransLab/network_runner.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+class NetworkRunner(NetworkRunnerBase):
+    def __init__(
+        self,
+        input_dir: Path,
+        output_dir: Path,
+        log_path: Path,
+        model_path: Path,
+        segmentron_args,
+    ):
+        root_path = os.path.abspath(os.path.dirname(__file__))
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/network_runner.py]: please rewrite using pathlib

File: detectors/TransLab/network_runner.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        sys.path.append(root_path)
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/network_runner.py]: pathlib

File: detectors/TransLab/network_runner.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        default_setup(args)
+
+        self.args = args
+        self.device = torch.device(args.device)
+
+        super().__init__(input_dir, output_dir, log_path, model_path)
+
+        self._init_dataset()
+
+    def _image_gen(self) -> Tuple[Any, Path, Tuple[int, int]]:
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/network_runner.py]: `Tuple[Any, Path, Tuple[int, int]]` looks scary. Maybe add a class that encapsulates this information

File: detectors/TransLab/network_runner.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        with torch.no_grad():
+            output, output_boundary = self.model.evaluate(img)
+
+            glass_res = output.argmax(1)[0].data.cpu().numpy().astype("uint8") * 127
+            glass_res = cv2.resize(glass_res, shape, interpolation=cv2.INTER_NEAREST)
+        return glass_res
+
+    def _write_img(self, img_name, prediction):
+        save_path = self.output_dir
+        cv2.imwrite(os.path.join(save_path, img_name), prediction)
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/network_runner.py]: pathlib

File: detectors/TransLab/network_runner.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+        self.val_loader = data.DataLoader(
+            dataset=val_dataset,
+            batch_sampler=val_batch_sampler,
+            num_workers=cfg.DATASET.WORKERS,
+            pin_memory=True,
+        )
+        self.classes = val_dataset.classes
+
+    def _set_batch_norm_attr(self, named_modules, attr, value):
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/network_runner.py]: Please use two underscores `__` before the name if a method is not expected to be overridden

File: detectors/TransLab/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,19 @@
+FROM pytorch/pytorch:2.0.0-cuda11.7-cudnn8-runtime
+
+ARG DETECTOR_DIR=TransLab
+
+WORKDIR /detector
+
+COPY ../common common
```
Comment (kirill-ivanov-a) [inline detectors/TransLab/Dockerfile]: Parent directory cannot be accessed in docker. [See](https://docs.docker.com/build/building/context/)

**Review by kirill-ivanov-a — CHANGES_REQUESTED** [review]

File: detectors/TransLab/network_runner.py (hunk 7 above)
Comment (misha-kis) [reply detectors/TransLab/network_runner.py]: I have not found a `pathlib` solution for adding directories to PATH, if this is what you ment

**Review by kirill-ivanov-a — APPROVED** [review]

**Review by anastasiia-kornilova — APPROVED** [review]

**Pull request merged**